### ISSUE TITLE
Virtual Actors

### ISSUE THREAD

**Issue by rogeralsing** [akkadotnet/akka.net#756]
This is not a suggestion for V1 but rather a general discussion around the topic of availability.

I have an idea on how to replicate the concept of Orleans virtual actors in Akka.NET.

Instead of trying to make our `ActorRef`always available, re-activatable.
We could base the concept on `ActorSelection` as actor selection is not tied to `UID` or any specific Actor incarnation.

And in order to make the actors always available, we could simply spin up requested actors when the `ActorSelection` is beeing processed in a system, that is when each part of the path is being traversed and `GetChild(string name)` is called.
If we have a special `VirtualActorCell` this cell could spin up actors for paths that does not exist right now.

Actors could also just stop themselves after x amount of idle time using `SerReceiveTimeout`

What is missing from this this is how to define a props for a given path. 
Orleans doesnt have that problem as they just request a given IGrainIface and the actor will become that interface.. which is super odd, as if the grain is requested two hours later with another interface, it will become an instance of that instead.. oh well :)

**Comment by Horusiath** [comment]
I'd like to hear some voices from Typesafe guys in this matter. What were their reasons for not to implement virtual actors in Akka?

**Comment by Aaronontheweb** [comment]
@Horusiath discussed this with @rkuhn this morning.

TL;DR; - Akka.ClusterSharding does this effectively - semantics are a little different but accomplishes the same goal. I'll leave it to @rkuhn to explain in more detail.

**Comment by rkuhn** [comment]
This is a good question: it illustrates some of the core design choices we have made when building and extending Akka. The foremost concern is always to crystallize the simplest consistent semantics that allow the expression of each feature, which is a convoluted way of saying “no magic”.

An `ActorRef` is a handle to an Actor that has been known to exist at some point in the past—meaning an `ActorRef` cannot be fabricated, it can only be obtained by creating that Actor. It has taken us some time to arrive at this conclusion, which is why `actorFor` is still in the process of being removed. The purpose of an `ActorRef` is to send messages to its Actor or to watch the lifecycle of that Actor; the latter only makes sense for Actors that have previously existed.

An `ActorSelection` is a possible Actor name (more precisely: `ActorPath`), which means that lifecycle monitoring does not work based on this information alone, only sending of messages is possible.

Both of these primitive constructs are fully self-contained, sending them around between Actors and ActorSystems is trivially possible by serializing them into URIs, which brings me to the current point under discussion: there is no place in either of these to contain information on which concrete Actor implementation type they shall point to. Therefore conjuring up an Actor within the framework in response to sending to either of these would be inconsistent as Roger already notes, letting pure chance decide which implementation to instantiate.

A principled solution to this problem is to set up an Actor that handles the creation of a certain class of other Actors. Then messages directed to this Actor will be redirected to an instance of the target type that either already exists or is created on demand. The added bonus is that this Actor creation is now hidden from the end user, explicitly taken out of their control, allowing the placement of the new instance to be decided by the framework. We call this [ClusterSharding](http://doc.akka.io/docs/akka/2.3.9/contrib/cluster-sharding.html), currently an experimental add-on that will move into the core distribution in Akka 2.4.0.

So, I hope this explains why and how we have implemented virtual Actors in a much more consistent fashion than Orleans.

**Comment by ReubenBond** [comment]
> What is missing from this this is how to define a props for a given path. 
Orleans doesnt have that problem as they just request a given IGrainIface and the actor will become that interface.. which is super odd, as if the grain is requested two hours later with another interface, it will become an instance of that instead.. oh well :)

I think you may have a slight misunderstanding, @rogeralsing: a grain is identified by the tuple of (Type, Id) - `GrainFactory.GetGrain<IOrganizationActor>(Guid.Empty)` and `GrainFactory.GetGrain<IUserActor>(Guid.Empty)` give me references to different actors.

**Comment by ReubenBond** [comment]
> So, I hope this explains why and how we have implemented virtual Actors in a much more consistent fashion than Orleans.

Maybe I misunderstood, @rkuhn, but doesn't your solution require explicitly sending every message through an intermediary? Would you mind posting a (pseudo) code snippet of how I could use virtual actors via Akka?

**Comment by rkuhn** [comment]
@ReubenBond Yes, you understand that correctly. It is the _Actor Way_ to model things as Actors, and here the concern is to distribute and lazily create Actors that is implemented that way. I understand where you are coming from, you think “let’s just enhance the ActorRef concept a little bit”, and we have tried things like that in the past and were always bitten by the semantic dilution: the Actor Model captures exactly the concept of Distribution, nothing less but also nothing more. Adding more concerns to the core pieces greatly reduces their power in sometimes non-obvious ways, they draw their strength from being extremely well-defined and minimal.

On a more practical note: what do you suppose should happen if I `watch()` a virtual `ActorRef`? What happens to “the Actor is created by its Parent” when everyone can create Actors by creating a certain `ActorSelection`? The core abstraction provides very clear identity and lifecycle semantics, let’s not muddy the waters by following some other ill-informed project.

Concerning a usage example, just follow the link I provided in my previous post.

**Comment by rkuhn** [comment]
> It's okay that these concepts don't map cleanly into actor frameworks, since virtual actors and actors are different models. There are merits to both and they can coexist happily.
> 
> Edit: I may have not been clear in this, but Virtual Actors are not a super set of Actors, they're quite different - as you've identified. The key difference (that virtual actors are eternal) has significant implications and many concepts from the Actor model do not map onto the Virtual Actor model.

I fully agree, which is why it is obvious to me that both should not share common types in an implementation: separate semantics need separate names. As such, “virtual Actors” might be a completely separate project or sub-project, which might or might not be built on top of regular Actors (as an implementation detail).

And to close the other loop: Akka’s ClusterSharding is not exactly the same as virtual Actors—it comes close in terms of support use-cases, but it lives within the confines of the Actor Model.

On a side note: I would really prefer that Orleans-style abstractions are called something else than Actors, because they are not the same thing. But we don’t always get to name all the things consistently ;-)

**Comment by ReubenBond** [comment]
I agree that a sub-project is likely the best approach if Akka users want Virtual Actors. Before Orleans was released, I was an Akka user and I wanted Virtual Actors, so perhaps there are others... on the other hand, they already have [Orbit](https://github.com/electronicarts/orbit) on the JVM and Orleans on the CLR.

> On a side note: I would really prefer that Orleans-style abstractions are called something else than Actors, because they are not the same thing. But we don’t always get to name all the things consistently ;-)

They're called Virtual Actors, sometimes Grains. They are quite similar to actors, so they are similarly named. The naming is consistent. There is a parallel with virtual memory & physical memory.

As an aside, I get the feeling that you're not convinced of the usefulness of Virtual Actors. Would you mind elucidating why for me?

**Comment by rkuhn** [comment]
> As an aside, I get the feeling that you're not convinced of the usefulness of Virtual Actors. Would you mind elucidating why for me?

The concept may make sense—I have not yet spent enough time pondering it to form a real opinion—but the implementation in Orleans is far too magical for my taste. Another way of putting this is that in the past I have learnt (the hard way) to eschew the “easy” and magical solutions in preference of actually simpler and more minimal abstractions. As an example, persistence is a core concern of how your business logic shall behave, it needs careful consideration in order to achieve the desired semantics (here the blue-sky scenario is completely uninteresting, it is the failure scenarios that pose all the problems). Another is whether you can tolerate split-brain scenarios or not, and what should be done when they happen. The third one is communication, which to my mind also must be explicit, and it must be explicitly different from local method calls for programmers to realize that “here be partial failure”.

I’m not saying that in Akka we have already succeeded in building the minimal solution to all problems that we consider to be in scope, but we are striving for the simple and non-magic everywhere, to the great benefit of our users.

**Comment by ReubenBond** [comment]
> The concept may make sense—I have not yet spent enough time pondering it to form a real opinion—but the implementation in Orleans is far too magical for my taste.

It's actually quite simple - there's really no magic. One valid concern is that it does too much and isn't modularised enough. That's fair, hopefully we can chop it up more. It would be nice if C# had traits like you get in Scala, it would make it easier to decompose the system.

> As an example, persistence is a core concern of how your business logic shall behave, it needs careful consideration in order to achieve the desired semantics (here the blue-sky scenario is completely uninteresting, it is the failure scenarios that pose all the problems).

Completely agree. People often get confused by Orleans' very simple pluggable persistence system. It's not essential or core to Orleans, but it makes things easier in the basic case. The Halo 4 team didn't use the in-built persistence, IIRC. [Caitie McCaffrey](https://twitter.com/Caitie) has given some talks on the saga: https://youtu.be/gY8zKZUazvo?t=1623.

Many Orleans' users use event sourcing for persistence, although there isn't yet a gold-standard alternative to `akka-persistence`. We use event sourcing, with snapshots provided by the in-built persistence system.

> Another is whether you can tolerate split-brain scenarios or not, and what should be done when they happen.

I've been considering implementing a CP directory for Orleans', but the in-built AP directory is currently sufficient for our purposes. As mentioned, Orbit apparently has a CP directory.

> The third one is communication, which to my mind also must be explicit, and it must be explicitly different from local method calls for programmers to realize that “here be partial failure”.

Orleans uses async/await, so communication and the potential for failures is made very apparent.

Play with Orleans at some point, poke through [the docs](https://dotnet.github.io/orleans/), or at least watch through [Caitie's video](https://youtu.be/gY8zKZUazvo?t=1623).

Akka and Orleans are both very mature projects (both around the same age), but they can both learn from each other and become better.

**Comment by rogeralsing** [comment]
I'm closing this issue, I guess we should really have feature discussions elsewhere and keep the issue tracker focused on issues.

**Comment by rkuhn** [comment]
@ReubenBond Playing around with Orleans is unlikely to happen since it seems to require VisualStudio—I have neither that nor the O/S to run it—but I am currently studying the documentation in great detail.

@rogeralsing I think this question deserves a properly informed response that has not yet been given. Where shall we continue?

**Comment by rogeralsing** [comment]
@rkuhn maybe there could be a generic repo to discuss Akka as a concept and overall api?
Platform agnostic?

**Comment by ReubenBond** [comment]
Thanks @rkuhn - I've got a VM for you with VS & Orleans if you want to play with code. No fuss, guaranteed! DM me on Twitter @reubenbond if you want details. Sorry for gunking up the pipes, @rogeralsing :smile:

**Comment by rogeralsing** [comment]
@rkuhn nodeschool has a discussion repo https://github.com/nodeschool/discussions

Something along those lines. Possibly under the akka organization
